### ISSUE TITLE
ccls not applying compile_commands.json arguments to header files

### ISSUE THREAD

**Issue by madscientist** [MaskRay/ccls#709]
### Observed behavior

I am installing a new version of ccls compiled from current HEAD master (sha 1dc69f45) built against clang 10.0.1.  I'm seeing basically the same behavior I reported in issue #335 although the cause is different: I've compiled a debug ccls and I don't get any crashes or warnings about invalid json, and I've deleted and rebuilt my compile_commands.json to be correct (I use CMake to generate it), and deleted pre-existing ccls cache etc.

When I edit a `.cpp` file and look at the verbose output log I see that all the correct command line arguments are obtained from compile_commands.json and things work.  But when I edit a `.h` file none of the command line arguments from compile_commands.json are used: only the options that appear in my `.ccls` file show up, and so parsing fails.

Full details below.

I'm happy to debug ccls if you can provide any pointers as to where to look and how to proceed.

### Expected behavior

I'm not sure what heuristic ccls uses to determine the right compiler options for a header file... does it find some .cpp file in the same directory maybe?  But anyway, _some_ set of compiler flags need to be used when parsing headers.

### Steps to reproduce

Delete .ccls-cache, here's my `.ccls` file:

```
%compile_commands.json
%h -x
%h c++-header
%h %hpp --include=Base.h
-Wno-pragma-once-outside-header
```

(I added the -x c++-header just to see if would help but it didn't make a difference)

Here's the "interesting" output in my ccls log file:

```
13:50:18 ccls           initialize.cc:297 I initializationOptions: {"compilationDatabaseCommand":"","compilationDatabaseDirectory":"","cache":{"directory":".ccls-cache","format":"binary","hierarchicalPath":false,"retainInMemory":2},"capabilities":{"documentOnTypeFormattingProvider":{"firstTriggerCharacter":"}","moreTriggerCharacter":[]},"foldingRangeProvider":true,"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}}},"clang":{"excludeArgs":[],"extraArgs":["--gcc-toolchain=/cc/x86_64-linux/cc/unknown","--sysroot=/cc/x86_64-linux/sysroot/rh75"],"pathMappings":[],"resourceDir":"/cc/x86_64-linux/ccls/clang"},"client":{"diagnosticsRelatedInformation":true,"hierarchicalDocumentSymbolSupport":true,"linkSupport":true,"snippetSupport":true},"codeLens":{"localVariables":true},"completion":{"caseSensitivity":2,"detailedLabel":true,"dropOldRequests":true,"duplicateOptional":true,"filterAndSort":true,"include":{"blacklist":[],"maxPathSize":30,"suffixWhitelist":[".h",".hpp",".hh",".inc"],"whitelist":[]},"maxNum":100,"placeholder":true},"diagnostics":{"blacklist":[],"onChange":1000,"onOpen":0,"onSave":0,"spellChecking":true,"whitelist":[]},"highlight":{"largeFileSize":2097152,"lsRanges":false,"blacklist":[],"whitelist":[]},"index":{"blacklist":[],"comments":2,"initialNoLinkage":false,"initialBlacklist":["."],"initialWhitelist":[],"maxInitializerLines":5,"multiVersion":0,"multiVersionBlacklist":[],"multiVersionWhitelist":[],"name":{"suppressUnwrittenScope":false},"onChange":false,"parametersInDeclarations":true,"threads":6,"trackDependency":2,"whitelist":[]},"request":{"timeout":5000},"session":{"maxNum":10},"workspaceSymbol":{"caseSensitivity":1,"maxNum":1000,"sort":true},"xref":{"maxNum":2000}}
13:50:18 ccls           initialize.cc:329 I use -resource-dir=/cc/x86_64-linux/ccls/clang
13:50:18 ccls             pipeline.cc:877 V(2) respond to RequestMessage: 13233
13:50:18 ccls           initialize.cc:361 I workspace folder: /src/
13:50:18 ccls              project.cc:426 I loaded /src/compile_commands.json
  ...
13:50:18 ccls              project.cc:284 I use /src/.ccls: %compile_commands.json %h -x %h c++-header %h %hpp --include=Base.h -Wno-pragma-once-outside-header
```
This it skips all the files in my compile_commands.json because I set everything to be initially blacklisted: the source is big enough that I don't like to index the entire thing every time.

Then the request for the file comes from my header:
```
13:50:18 ccls         sema_manager.cc:745 I clear all sessions
13:50:18 preamble     sema_manager.cc:736 I create session for /src/Client/ClientUtils.h
  -x c++-header --include=Cloud.h -Wno-pragma-once-outside-header /src/Client/ClientUtils.h --gcc-toolchain=/cc/x86_64-linux/cc/unknown --sysroot=/cc/x86_64-linux/sysroot/rh75 -working-directory=/src/
13:50:18 ccls             pipeline.cc:493 I loaded project. Refresh semantic highlight for all working file.
13:50:18 ccls             pipeline.cc:493 I loaded project. Refresh semantic highlight for all working file.
13:50:18 indexer1         pipeline.cc:386 I parse /src/Client/ClientUtils.h error:1 'Base.h' file not found
  -x c++-header --include=Cloud.h -Wno-pragma-once-outside-header /src/Client/ClientUtils.h --gcc-toolchain=/cc/x8
6_64-linux/cc/unknown --sysroot=/cc/x86_64-linux/sysroot/rh75 -working-directory=/src/
13:50:18 indexer1         pipeline.cc:397 I store index for /src/Client/ClientUtils.h (delta: 0)
```

You can see that the set of options provided contain both the contents of `.ccls` and the extra options the LSP client provided, but they don't contain any options from compile_commands.json.

On the other hand if I visit a `.cpp` file, I get everything I expect:

```
13:50:20 indexer2         pipeline.cc:386 I parse /src/Client/ClientUtils.cpp
  /cc/x86_64-linux/bin/x86_64-rh75-linux-gnu-g++ --driver-mode=g++ -I/src/Foundation -std=c++17 -ggdb3 -fPIC -O2 -march=nehalem -mtune=intel -fno-omit-frame-pointer -ffile-prefix-map=/src/DB-29265/= -pthread -Wall -Werror -Wno-unused-function -Wno-strict-aliasing -Wformat-security -Wtype-limits -Wunused-but-set-parameter -Wlogical-op -Wignored-qualifiers -Wformat-signedness -Winvalid-pch -o CMakeFiles/ClientBase.dir/ClientUtils.cpp.o -c /src/Client/ClientDateUtils.cpp -Wno-pragma-once-outside-header --gcc-toolchain=/cc/x86_64-linux/cc/unknown --sysroot=/cc/x86_64-linux/sysroot/rh75 -working-directory=/src/Client
```

and things parse properly.

### System information

* ccls version (`git describe --tags --long`): ccls version 0.20190823.6-27-g1dc69f45
* clang version: clang version 10.0.1 (https://github.com/llvm/llvm-project.git ef32c611aa214dea855364efd7ba451ec5ec3f74)
* OS: Ubuntu 20.04 (but I compiled all the things myself)
* Editor: Emacs 26.3
* Language client (and version): lsp-mode 20201021.1658 / ccls 20200820.308 (from melpa)

My ccls-initialization-options in Emacs are set as:

```
  (setq ccls-initialization-options
        (append (:clang (:extraArgs ("--gcc-toolchain=/cc/x86_64-linux/cc/unknown" "--sysroot=/cc/x86_64-linux/sysroot/rh75")))
                `(:index (:threads 6
                          :initialBlacklist (".")
                          )))
        ;; Enable for debugging
        ccls-args '("--log-file=ccls.out" "-v=10")
        )
```

**Comment by madscientist** [comment]
Just to note, my previous version of ccls 0.20190823.5-9-gcef02034 built from sha cef02034 still works fine...

**Comment by madscientist** [comment]
Well I can't find that sha anywhere so I don't know what happened, unless maybe there was a force-push of the Git history?  Anyway, the version I have that works was built around a year ago, on or about Nov 28 2019 so certainly nothing LATER than that is included.

**Comment by madscientist** [comment]
As I suppose might be expected from the behavior vs. the commit description, bisecting shows the failure starts with ef8473ecc0b

Something in here doesn't play well with header files... at least not in my configuration.  I'll try to debug further since it's a simple patch.

**Comment by madscientist** [comment]
So here's what happens: in findEntry() it's invoked with path set to `/src/Client/ClientDate.h`.  We're looking for the best match and we walk through rooot2folder.  We find the best fit, then we enter the if-statement (because match is false by default) and line 521.

In that if statement, we search for path in folder.path2entry_index but we don't find it (I assume because header files are not present in this list).  As a result, we don't set exact_match _and_ we don't set `best_compdb_folder` or `best`.

Then we drop down to the changed if-statement at line 536, which is this diff:
```
-  if (best_dot_ccls_args && !(append = appendToCDB(*best_dot_ccls_args)) &&
-      !exact_match) {
+  if (best_dot_ccls_args &&
+      (!best_compdb_folder ||
+       (!(append = appendToCDB(*best_dot_ccls_args)) && !exact_match))) {
```
Here, we've added a check for `best_compdb_folder` and if it's zero (which is is for header files) then we _short-circuit_ the call to `appendToCDB()` and we enter the then-part of the if statement, which is only supposed to be entered if we don't have a compile_commands.json file... hence the missing arguments.

I don't understand why this change was made, but it's not correct as written...

**Comment by MaskRay** [comment]
> As I suppose might be expected from the behavior vs. the commit description, bisecting shows the failure starts with ef8473e
> Something in here doesn't play well with header files... at least not in my configuration. I'll try to debug further since it's a simple patch.

It was indeed caused by ef8473e, which was intended to make recursive `.ccls` files work when there is no `compile_commands.json`. It messed up the case for `%compile_commands.json`.

> Well I can't find that sha anywhere so I don't know what happened, unless maybe there was a force-push of the Git history?

Apologies for that. There were force-pushes of the Git history. It is very bad in collaborating but works well when you want to keep clean history and collaboration is very low...

I believe the patch can fix the problem, 

```patch
--- a/src/project.cc
+++ b/src/project.cc
@@ -522,9 +522,10 @@ Project::Entry Project::findEntry(const std::string &path, bool can_redirect,
         auto it = folder.path2entry_index.find(path);
         if (it != folder.path2entry_index.end()) {
           Project::Entry &entry = folder.entries[it->second];
-          exact_match = entry.filename == path;
+          // best->compdb_size is >0 for an explicit compile_commands.json
+          // entry, 0 for an implicit entry.
+          exact_match = entry.compdb_size > 0 && entry.filename == path;
           if ((match = exact_match || can_redirect) && entry.compdb_size) {
-            // best->compdb_size is >0 for a compdb entry, 0 for a .ccls entry.
             best_compdb_folder = &folder;
             best = &entry;
           }
@@ -534,8 +535,7 @@ Project::Entry Project::findEntry(const std::string &path, bool can_redirect,
 
   bool append = false;
   if (best_dot_ccls_args &&
-      (!best_compdb_folder ||
-       (!(append = appendToCDB(*best_dot_ccls_args)) && !exact_match))) {
+      !(append = appendToCDB(*best_dot_ccls_args)) && !exact_match) {
     // If the first line is not %compile_commands.json, override the compdb
     // match if it is not an exact match.
     ret.root = ret.directory = best_dot_ccls_root;
```

**Comment by madscientist** [comment]
Force push doesn't really bother me: I'm not a militant on the topic.  Just wasn't sure if I'd built my ccls from some other fork or something.

Thanks for the patch; I applied it and it does seem to work correctly for me.  Cheers!

**Comment by MaskRay** [comment]
Thanks for the report! Fixed by e3e687f81d60f4a4b12aa0e44bb12efda811b4b1
